design/ratelimit-design.md

```diff
@@ -1,175 +1,502 @@
+    # on the Envoy local rate limiter.
+    # See ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket.
+    unit: second
+    # burst defines how many additional requests above the baseline requests
+    # are allowed in a short period of time. This, along with "requests", 
+    # programs the "max_tokens" field on the Envoy local rate limit filter.
+    # See ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket.
+    burst: 20
+  # global defines global rate limiting properties for the virtual host or route.
+  global:
+    # rateLimitDescriptors defines the descriptors to be generated and sent to the
+    # external Rate Limiting Service (RLS) for a rate limit decision.
+    rateLimitDescriptors:
+    # This descriptor is generated only if the x-steve-ratelimit header
+    # is present on the request.
+    - items:
+        # adds a descriptor entry with a key of "generic_key" and a value
+        # of "s1".
+        - genericKey:
+            value: s1
+
+        # adds a descriptor entry with a key of "remote_address" and a value
+        # equal to the trusted address from x-forwarded-for.
+        - remoteAddress: {}
+
+        # adds a descriptor entry with a key of "steve-ratelimit"
+        # and a value equal to the value of the "x-steve-ratelimit" header.
+        - requestHeader:
+            descriptorKey: steve-ratelimit
+            headerName: x-steve-ratelimit
+        
+    # This descriptor is always generated since it's just a static value.
+    - items:
+        - genericKey:
+            value: s1
+```
 
-___Note: Much of the following documentation is from Lyft's implementation [github repo](https://github.com/lyft/ratelimit), but copied here for completeness of this design document.___
+This API maps closely to Envoy's rate limiting API.
 
-The rate limit service is a Go/gRPC service designed to enable generic rate limit scenarios from different types of applications. Applications request a rate limit decision based on a domain and a set of descriptors. The service reads the configuration from disk via runtime, composes a cache key, and talks to the Redis cache. A decision is then returned to the caller.
+For local rate limiting, the rate limit itself is defined inline in the `RateLimitPolicy`, as token bucket properties.
+Since local rate limits apply *per Envoy*, each pod in the Envoy daemon set will have a token bucket with these properties.
 
-New arguments to Contour:
+For global rate limiting, per Envoy's API, the `RateLimitPolicy` only defines a list of descriptors to send to an external RLS for a given request.
+All [descriptor entries defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action) will be supported except `metadata` and `dynamic_metadata`.
```

Comment:
nit: Maybe define what these all would be in the API? For example `requestHeaders` can also have `skip_if_absent` which isn't in the examples. 

Also, I assume we're not going to support `SourceCluster` & `DestinationCluster` which might be new since this design doc was drafted.

Reply:
OK I'll spec out the details for all of them.

Reply:
ok to merge as-is, can do this work during implementation phase.